_posts/2018-02-06-apogee-technical-retrospective.markdown

```diff
@@ -0,0 +1,262 @@
+
+Apogee had some really interesting technical challenges that I enjoyed solving, and I'm excited to share some lessons I learned. So let's dive in!
+
+<!-- more -->
+
+We built a prototype as a "pure" Add-on, written only inside Google's sandbox, but that approach wouldn't work for us in production: the Add-on environment was just too difficult to work with. Google expects you to write Add-ons in their in-browser [Script Editor][scripts] and – while whether or not that editor is _good_ is a matter of preference – the environment isn't suited for collaborating or unit testing. Additionally, we could not get Add-on deploys automated, so we'd like to minimize how often we _have_ to deploy.
+
+So we split things up. Instead of building all Apogee's logic into an Add-on, we decided to build two pieces: a very thin Add-on and a Rails server with all the real logic.
+
+(Because Apogee necessarily includes information about how our partners format their data, we decided not to open source it. Data formats are _probably_ not sensitive, but that's a judgement best left up to our partners.)
+
+## Apogee Add-on
+
+The Add-on we built is very simple, by design. Our goal was to make an Add-on that was flexible enough such that we would need to deploy it less frequently than adding new parsers.
+
+Add-on responsibilities include:
+
+- fetching the available parsers from the server.
+- setting up an Add-on user interface (a menu of partners, each with available parsers).
+- responding to invocations from that interface. 
+
+Based on the parser selected by the user, Apogee gathers the required data from the current spreadsheet, sends it to the server for processing, and appends the results to the sheet. Pretty straightforward, you'd think.
+
+Unfortunately, Google Add-ons are a bit... strange. The Add-on itself is executed in Google's datacentres (not the user's browser) and is written in [JavaScript 1.6-ish][version]. The execution environment also lacks an event loop which, again, is a bit strange.
+
+Rather than deal with a weird version of JavaScript, we decided to write the Add-on in [TypeScript][] and compile down to something Google can execute. We also found [open source typings][typings] for the Google APIs, which helped a lot. Google also provides access to certain whitelisted libraries, including [Lodash][], which is handy.
+
+Add-ons also have a somewhat complex permissions and authentication model. The [documentation][] provided is a great illustration of why _complete_ documentation is not necessarily _effective_ documentation. If you already understand what you're doing, the docs are a good reference, but I found them difficult to learn from.
```

Comment:
LOL! Well said.

Reply:
Docs are like tests, you need multiple levels: https://twitter.com/kosamari/status/852319140060823553